### PR TITLE
Adding hexbin to vis

### PR DESCRIPTION
**Pull request by dvzacharycutler**
Closes https://github.com/datavisyn/reprovisyn/issues/253

### Developer Checklist (Definition of Done)

**Issue**

- [x] All acceptance criteria from the issue are met
- [x] Tested in latest Chrome/Firefox

**UI/UX/Vis**

- [x] Requires UI/UX/Vis review
  - [ ] Reviewer(s) are notified (_tag assignees_)
  - [ ] Review has occurred (_link to notes_)
  - [ ] Feedback is included in this PR
  - [ ] Reviewer(s) approve of concept and design

**Code**

- [x] Branch is up-to-date with the branch to be merged with, i.e., develop
- [ ] Code is cleaned up and formatted
- [ ] Unit tests are written (frontend/backend if applicable)
- [ ] Integration tests are written (if applicable)

**PR**

- [x] Descriptive title for this pull request is provided (will be used for release notes later)
- [x] Reviewer and assignees are defined
- [ ] Add type label (e.g., *bug*, *feature*) to this pull request
- [ ] Add release label (e.g., `release: minor`) to this PR following [semver](https://semver.org/)
- [x] The PR is connected to the corresponding issue (via `Closes #...`)
- [x] [Summary of changes](#summary-of-changes) is written

### Summary of changes

- Adding back the hexbin plot now that d3 is compatible. Also includes small changes to the selection logic which drastically improves speed for large selections. Also made the vis component import dynamically in the ARankingView, so as not to always import mantine in tdp_core, but importantly i HAVE NOT tested this yet, as reprovisyn doesn't actually use the component. Needs to be tested in an application that uses the Vis component via the ARankingView.

### Screenshots

[newHexin.webm](https://user-images.githubusercontent.com/87644761/191507893-0f01bf73-9130-496f-9d81-597f569748d9.webm)

### Additional notes for the reviewer(s)

-  TODO:: Test the Vis component via ARankingView, as mentioned in summary of changes
Thanks for creating this pull request 🤗

### CODE REVIEW

File: src/vis/hexbin/Hexplot.tsx [hunk 1]
```diff
@@ -0,0 +1,452 @@
...
+
+        selectionCallback(allSelectedPoints);
+
+        d3v7.select(this).call(brush.move, null);
+      }),
+    );
+  }, [width, height, id, hexes, selectionCallback, config.dragMode, xZoomTransform, yZoomTransform, zoomScale, xScale, yScale, margin]);
+
+  // TODO: svg elements seem weird with style/classNames. I can directly put on a transform to a g, for example, but it seems to work
+  // differently than if i use style to do so
```
Comment (thinkh) [inline src/vis/hexbin/Hexplot.tsx:392]: Sorry, I don't understand.

File: src/vis/hexbin/PieChart.tsx [hunk 2]
```diff
@@ -0,0 +1,34 @@
...
+    return d3v7.arc().innerRadius(0).outerRadius(radius);
+  }, [radius]);
+
+  const id = React.useMemo(() => uniqueId('PieNum'), []);
+
+  return (
+    <g style={{ transform }}>
+      {pie(data).map((slice, i) => {
+        // TODO: Why are indexes bad in the key? how else to do this? Also, I think the typings for arc are wrong, which is why im typing slice to any
+        // eslint-disable-next-line react/no-array-index-key
```
Comment (thinkh) [inline src/vis/hexbin/PieChart.tsx:29]: See https://blog.logrocket.com/12-essential-eslint-rules-react/#react-no-array-index-key and https://github.com/jsx-eslint/eslint-plugin-react/blob/master/docs/rules/no-array-index-key.md.

**Review by thinkh — CHANGES_REQUESTED** [review]
Awesome new visualization. It looks already very good. Please see my comments below.

- [ ] The mouse while should work in 

If no numerical columns are selected the message currently takes the full height of the vies view (should be only the height of the text). 

image

- [x] Hexbin plot (sentence case)
- [x] Invalid settings (sentence case)
- [x] Convert invalid settings card to [Mantine info alert](https://mantine.dev/core/alert/) (currently bootstrap)
   - Also for all other cards and visualizations
- [x] Horizontally center the info alert
- [x] Hexbin options (sentence case)
   image
- [x] Add tooltips to brush option buttons
   image
- [ ] Enable zoom with mouse wheel in brush mode (currently, it only works in pan mode)
- [ ] Make list of categories in legend scrollable
   image

File: src/vis/hexbin/Hexplot.tsx [hunk 3]
```diff
@@ -0,0 +1,452 @@
...
+    // TODO: Im cheating a bit here by appending the id/color value to each hex, breaking the types.
+    // is there a better way to type this?
+    return d3Hexbin(inputForHexbin) as unknown as HexbinBin<[number, number, string, string]>[];
+  }, [currentColorColumn, currentX, d3Hexbin, xScale, yScale, currentY]);
+
+  // simple radius scale for the hexes
+  const radiusScale = useMemo(() => {
+    if (colsStatus === 'success') {
+      const min = d3v7.min(hexes.map((h) => h.length));
+      const max = d3v7.max(hexes.map((h) => h.length));
```
[the remark this answers is not in the document]
Comment (dvzacharycutler) [reply src/vis/hexbin/Hexplot.tsx:241]: done

File: src/vis/hexbin/Hexplot.tsx [hunk 4]
```diff
@@ -0,0 +1,452 @@
...
+    }
+
+    return null;
+  }, [colsStatus, hexes, config.hexRadius]);
+
+  // simple opacity scale for the hexes
+  const opacityScale = useMemo(() => {
+    if (colsStatus === 'success') {
+      const min = d3v7.min(hexes.map((h) => h.length));
+      const max = d3v7.max(hexes.map((h) => h.length));
```
[the remark this answers is not in the document]
Comment (dvzacharycutler) [reply src/vis/hexbin/Hexplot.tsx:256]: done

File: src/vis/hexbin/HexbinVisSidebar.tsx [hunk 5]
```diff
@@ -0,0 +1,83 @@
...
+            <HexbinOptionSelect callback={(hexbinOptions: EHexbinOptions) => setConfig({ ...config, hexbinOptions })} currentSelected={config.hexbinOptions} />
+          ) : null}
+        </Stack>
+        <Divider my="sm" />
+        <Stack>
+          <HexSizeSlider currentValue={config.hexRadius} callback={(hexRadius: number) => setConfig({ ...config, hexRadius })} />
+          <HexSizeSwitch currentValue={config.isSizeScale} callback={(isSizeScale: boolean) => setConfig({ ...config, isSizeScale })} />
+          <HexOpacitySwitch currentValue={config.isOpacityScale} callback={(isOpacityScale: boolean) => setConfig({ ...config, isOpacityScale })} />
+        </Stack>
+        {/* <HexbinOptionSelect callback={(hexbinOptions: EHexbinOptions) => setConfig({ ...config, hexbinOptions })} currentSelected={config.hexbinOptions} /> */}
```
[the remark this answers is not in the document]
Comment (dvzacharycutler) [reply src/vis/hexbin/HexbinVisSidebar.tsx:77]: done

File: src/vis/sidebar/HexbinOptionSelect.tsx [hunk 6]
```diff
@@ -0,0 +1,17 @@
...
+  currentSelected: EHexbinOptions;
+}
+
+export function HexbinOptionSelect({ callback, currentSelected }: HexbinOptionSelectProps) {
+  const options = [
+    { value: EHexbinOptions.COLOR, label: EHexbinOptions.COLOR },
+    { value: EHexbinOptions.BINS, label: EHexbinOptions.BINS },
+    { value: EHexbinOptions.PIE, label: EHexbinOptions.PIE },
+  ];
+  return <Select label="Hexbin Options" onChange={(e) => callback(e as EHexbinOptions)} data={options} value={currentSelected} />;
```
[the remark this answers is not in the document]
Comment (dvzacharycutler) [reply src/vis/sidebar/HexbinOptionSelect.tsx]: done

**Review by thinkh — CHANGES_REQUESTED** [review]
@dvzacharycutler Thanks for your bug fixes. It works now much better. I still found a few inconsistencies that should be removed before merging this PR.

### Steps to reproduce

1. Select hexbin plot
2. Remove all columns

### Observed behavior

The message shows `Scatterplot`.

grafik

### Expected behavior

1. The message should mention the _Hexbin plot_ when this plot is selected
2. Consistent spelling: _Scatter plot_ vs. _Scatterplot_

**Review by thinkh — CHANGES_REQUESTED** [review]
@dvzacharycutler I tested this branch with Ordino public in a local workspace and found a couple of problems. The problems might be related to  #786 and not specifically to this PR. You can decide where to fix it.

Here an overview of the interaction with the general visualization of this branch:

[ordino-hexbin.webm](https://user-images.githubusercontent.com/5851088/200290986-db662af6-4482-458e-b8d6-f3a8a97c90bf.webm)

### Scatter plot: Opacity does not have an effect

Sometimes the opacity slider does not have any effect.
 
ordino-scatter-opacity-0

ordino-scatter-opacity-1

As you can see in the video above, it worked somehow after playing around and switching visualization. But it does not work when I opened it the first time (see screenshots).

### Scatter plot: Legend overlaying visualization

The legend is overlaying the visualization and makes it impossible to use. Could be a flex problem in Ordino?

ordino-scatter-overplotting

### Hexbin plot: Wrong options panel position

When switching to hexbin plot the options panel is positioned absolute to the window and not the visualization container. (This does not happen for the other visualizations.) After closing and opening the panel again it is positioned correctly. 

Furthermore the hexbin plot is empty which should be the case because the scatter plot shows data.

ordino-broken-hexbin

### Various plots: scroll position and options panel

For all plot types the visualization is positioned far to the right which requires to scroll to the right. Then the options panel is positioned somewhere in the center.

ordino-stripplot-scrolling

ordino-stripplot-scrolling-options-open

**Comment by thinkh** [comment]
In addition, I saw now that the options panel is also overlaying the Ordino start menu.

image

**Review by thinkh — APPROVED** [review]
Thanks for the fixes. Looks good to me.

**Pull request merged**